warehouse/query-core/src/main/java/datawave/query/iterator/CachingIterator.java

```diff
@@ -0,0 +1,46 @@
+package datawave.query.iterator;
+
+import com.google.common.collect.PeekingIterator;
+
+import java.util.Iterator;
+
+public class CachingIterator<T> implements PeekingIterator<T> {
+    private Iterator<T> delegate;
+    private T next;
+    
+    public CachingIterator(Iterator<T> delegate) {
+        this.delegate = delegate;
+        
+        if (delegate.hasNext()) {
+            next = delegate.next();
+        }
+    }
+    
+    @Override
+    public T peek() {
+        return next;
+    }
+    
+    @Override
+    public boolean hasNext() {
+        return next != null;
+    }
+    
+    @Override
+    public T next() {
+        T toReturn = next;
+        
+        if (delegate.hasNext()) {
+            next = delegate.next();
+        } else {
+            next = null;
+        }
+        
+        return toReturn;
+    }
+    
+    @Override
+    public void remove() {
+        // no-op
```

Comment:
Is there a reason you wouldn't delegate this method to the delegate iterator? Looks like PeekingIterator will throw an UnsupportedOperationException here, so we've turned that into a no-op. Is that desired?

Reply:
good point